### PR TITLE
Race condition fixes for OAuth flow

### PR DESCRIPTION
**Pull request by elorzafe**
_Issue #, if available:_
Fixes: #5133 #4992 #4988 #4385 #3841
_Description of changes:_
Awaiting in progress OAuth signIn flow for `currentUserPoolUser` until `cognitoHostedUI` event or timeout.

Tested in plain JS App, react-native and react-native with expo
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/Auth.ts [hunk 1]
```diff
@@ -1104,8 +1107,26 @@ export class AuthClass {
...
-				.then(() => {
-					const user = that.userPool.getCurrentUser();
+				.then(async () => {
+					let user = that.userPool.getCurrentUser();
+
+					if (this.isOAuthInProgress(user)) {
+						await new Promise((res, rej) => {
+							const timeoutId = setTimeout(rej, OAUTH_FLOW_MS_TIMEOUT);
+
+							Hub.listen('auth', ({ payload }) => {
```
Comment (manueliglesias) [inline packages/auth/src/Auth.ts:1117]: Store reference to listener so you can remove it later (in both the reject and resolve cases)

File: packages/auth/src/Auth.ts [hunk 2]
```diff
@@ -68,6 +68,7 @@ import { AuthErrorTypes } from './types/Auth';
 
 const logger = new Logger('AuthClass');
 const USER_ADMIN_SCOPE = 'aws.cognito.signin.user.admin';
+const OAUTH_FLOW_MS_TIMEOUT = 30000;
```
Comment (manueliglesias) [inline packages/auth/src/Auth.ts:71]: I'd say lets change this to 10 secs as per https://www.nngroup.com/articles/response-times-3-important-limits/

File: packages/auth/__tests__/auth-unit-test.ts [hunk 3]
```diff
@@ -3267,6 +3269,67 @@ describe('auth unit test', () => {
...
+
+			expect.assertions(2);
+
+			MemoryStorage.setItem('amplify-redirected-from-hosted-ui', true);
+
+			auth.currentUserPoolUser().then(resUser => {
+				expect(resUser).toEqual(user);
+				expect(spyon).toBeCalledTimes(2);
+				MemoryStorage.removeItem('amplify-redirected-from-hosted-ui');
+				spyon.mockClear();
```
Comment (manueliglesias) [inline packages/auth/__tests__/auth-unit-test.ts:3323]: What about doing e.g.

```javascript
  jest.resetModules();
  jest.resetAllMocks();
```

on a `beforeEach(() => {})`

File: packages/auth/src/Auth.ts [hunk 4]
```diff
@@ -1177,6 +1198,13 @@ export class AuthClass {
 		});
 	}
 
+	private isOAuthInProgress(user?: CognitoUser): boolean {
+		const fromHostedUI = this._storage.getItem(
+			'amplify-redirected-from-hosted-ui'
+		);
+		return !user && fromHostedUI;
```
Comment (manueliglesias) [inline packages/auth/src/Auth.ts:1205]: `fromHostedUI === 'true'`

File: packages/auth/src/Auth.ts [hunk 5]
```diff
@@ -1104,8 +1107,26 @@ export class AuthClass {
...
+					let user = that.userPool.getCurrentUser();
+
+					if (this.isOAuthInProgress(user)) {
+						await new Promise((res, rej) => {
+							const timeoutId = setTimeout(rej, OAUTH_FLOW_MS_TIMEOUT);
+
+							Hub.listen('auth', ({ payload }) => {
+								const { event } = payload;
+
+								if (event === 'cognitoHostedUI') {
```
Comment (ericclemmons) [inline packages/auth/src/Auth.ts:1120]: Sure enough!  

Screen Shot 2020-05-28 at 10 58 09 AM

File: packages/auth/src/Auth.ts [hunk 6]
```diff
@@ -1104,8 +1107,26 @@ export class AuthClass {
 		const that = this;
```
[the remark this answers is not in the document]
Comment (elorzafe) [reply packages/auth/src/Auth.ts:1107]: Yeah, but this need a refactor

File: packages/auth/__tests__/auth-unit-test.ts [hunk 7]
```diff
@@ -3267,6 +3269,67 @@ describe('auth unit test', () => {
 			spyon4.mockClear();
 			spyon5.mockClear();
 		});
+
+		test.skip('hosted UI in progress, signIn success', done => {
```
[the remark this answers is not in the document]
Comment (elorzafe) [reply packages/auth/__tests__/auth-unit-test.ts:3273]: The problem with this is that is working with `test.only` only, I am looking with some of the other tests are breaking this.

File: packages/auth/src/Auth.ts (hunk 5 above)
Comment (elorzafe) [reply packages/auth/src/Auth.ts:1120]: Thanks @ericclemmons for the feedback :) . I am working on Manuel comments, so I will update the PR.

**Comment by elorzafe** [comment]
> Overall, looks good to me. I'll wait on some of the comment updates to approve.

Thanks @jordanranz  I will update this pr with the comments soon

File: packages/auth/src/Auth.ts [hunk 8]
```diff
@@ -1101,11 +1105,44 @@ export class AuthClass {
 		if (!this.userPool) {
 			return this.rejectNoUserPool();
 		}
-		const that = this;
+
 		return new Promise((res, rej) => {
 			this._storageSync
-				.then(() => {
-					const user = that.userPool.getCurrentUser();
+				.then(async () => {
```
Comment (jordanranz) [inline packages/auth/src/Auth.ts:1111]: Could we possibly move this anon function into it's own function? It is getting quite large to be nested.

**Pull request merged**